samza-api/src/main/java/org/apache/samza/table/RecordDoesNotExistException.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table;
+
+/**
+ * Custom exception which can be thrown by implementations of {@link org.apache.samza.table.remote.TableWriteFunction}
+ * when {@link AsyncReadWriteTable#updateAsync(Object, Object, Object...)} fails due an existing record not being
+ * present for the given key. {@link org.apache.samza.operators.MessageStream#sendUpdateTo(Table, Object...)} will
+ * attempt to call {@link AsyncReadWriteTable#putAsync(Object, Object, Object...)} instead to insert a new record if a
+ * default is provided.
+ */
+public class RecordDoesNotExistException extends RuntimeException {
```

Comment:
Is it possible to add a test to test out the use case when this exception happens? It will assure we will do the put first with default value then post again.

Reply:
I did add a test case  in `doTestStreamTableJoinRemoteTableWithFirstTimeUpdates` in `TestRemoteTableEndToEnd`. The TableWriteFunction maintains a map of records and throws a `RecordNotFoundException` for first time updates. Let me revisit it to see if I missed out of any other scenarios.

Reply:
Cool. If there is a test then it is fine. I was searching for this exception in the test cases but I didn't find it. I guess it might be because too many files are in this pr. My suggestion is that next time let's separate large prs into smaller ones. For example, this one can break into the core changes + batching changes.